File: python/geometrickinematicscalc.py
import math

def radToDegree(radians):
  return radians * (180/math.pi)

def degToRadian (degrees):
  return degrees * (math.pi / 180)

def forward(q1d,q2d,l1,l2):
  q1 = degToRadian(q1d)
  q2 = degToRadian(q2d)
  points = [l1 * math.cos(q1) + l2 * math.cos(q1 + q2), l1 * math.sin(q1) + l2 * math.sin(q1 + q2)]
  return points

def forward2(q1d, q2d, l1, l2, offset_q1):
    q1 = degToRadian(q1d) + degToRadian(offset_q1)
    q2 = degToRadian(q2d)
    points = [l1 * math.cos(q1) + l2 * math.cos(q1 + q2), l1 * math.sin(q1) + l2 * math.sin(q1 + q2)]
    return points


def inverse(x,y,l1,l2):
  r1 = math.sqrt(pow(x,2) + pow(y,2))
  m = (pow(r1,2) - pow(l1,2) - pow(l2,2)) / (2 * l1 * l2)
  b1 = math.atan2(y,x)
  m2 = (pow(r1,2) + pow(l1,2) - pow(l2,2)) / (2 * r1 * l1)
  angles = [radToDegree(b1 - math.acos(m2)), radToDegree(math.acos(m))]
  return angles

def inverse2(x, y, l1, l2, offset_q1):
    r1 = math.sqrt(pow(x, 2) + pow(y, 2))
    m = (pow(r1, 2) - pow(l1, 2) - pow(l2, 2)) / (2 * l1 * l2)
    b1 = math.atan2(y, x) + degToRadian(offset_q1)
    m2 = (pow(r1, 2) + pow(l1, 2) - pow(l2, 2)) / (2 * r1 * l1)
    angles = [radToDegree(b1 - math.acos(m2)), radToDegree(math.acos(m))]
    return angles


temp = forward(178,160,106,135)
print(temp[0])
print(temp[1])
temp = inverse(80,213,106,135)
print(temp[0])
print(temp[1])


{"shoulder_offset": 170.57635653800241, "elbow_offset": -165.8135947109648}

File: cpp/geometrickinematicscalc.cpp
#include <iostream>
#include <vector>
#include <cmath>

const long double pi = atan(1)*4;

double radToDegree (double radians) {
  return radians * (180/pi);
}

double degToRadian (double degrees) {
  return degrees * (pi / 180);
}

std::vector<double> forward(double q1d, double q2d, double l1, double l2) {
  double q1 = degToRadian(q1d);
  double q2 = degToRadian(q2d);
  std::vector<double> result;
  result.push_back(l1 * cos(q1) + l2 * cos(q1 + q2));
  result.push_back(l1 * sin(q1) + l2 * sin(q1 + q2));
  return result;
}

std::vector<double> inverse(double x, double y, double l1, double l2) {
  std::vector<double> result;
  double r1 = sqrt(pow(x, 2) + pow(y, 2));
  double m = (pow(r1, 2) - pow(l1, 2) - pow(l2, 2)) / (2 * l1 * l2);
  double b1 = atan2(y, x);
  double m2 = (pow(r1, 2) + pow(l1, 2) - pow(l2, 2)) / (2 * r1 * l1);
  double q1 = b1 - acos(m2);
  double q2 = acos(m);
  result.push_back(radToDegree(q1));
  result.push_back(radToDegree(q2));
  return result;
}

int main(void) {
  std::vector<double> temp;

  double offset_q1 = 10.0;  // Offset para q1 en grados
  double offset_q2 = -5.0;  // Offset para q2 en grados

  temp = forward(0, -200, 106, 135); // points to angles
  std::cout << "directa:\n";
  std::cout << temp.at(0) + 165<< "\n";
  std::cout << temp.at(1) - 170 << "\n";

  temp = inverse(180 + 165, 180 - 170, 106, 135); // angles to points
  std::cout << "inversa:\n";
  std::cout << temp.at(0) << "\n";
  std::cout << temp.at(1) << "\n";

  return 0;
}

